### ISSUE TITLE
JSImport/Export in Maui Blazor Hybrid

### ISSUE THREAD

**Issue by hawkerm** [dotnet/AspNetCore.Docs#31331]
### Description

I find this article's examples a bit hard to process, it's not clear exactly what goes where. But following it and trying to run it in my Blazor Maui Hybrid app, I get a "System.Runtime.InteropServices.JavaScript is not supported on this platform." error on the JSHost.ImportAsync line... which I don't understand...

How do I interop with JavaScript in a Blazor Maui Hybrid app? With all the different flavors of Blazor/ASP/MAUI Hybrid configurations out there, it's very difficult to find the right proper documentation or know if the documentation applies to the specific setup of my project.

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/import-export-interop?view=aspnetcore-8.0

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/javascript-interoperability/import-export-interop.md

### Document ID

8aab92a0-c05f-ef01-fb74-a0bf525aae69

### Article author

@guardrex

**Comment by hawkerm** [comment]
Ugh, looks like this isn't a supported scenario? https://github.com/dotnet/maui/issues/13551 - though then I don't understand what the equivalent is...

**Comment by guardrex** [comment]
Hello @hawkerm ... Yes, I was just typing that .........

As the article states ...

> This article describes an alternative JS interop approach specific to client-side components executed on WebAssembly. 

... so this article's scenarios aren't appropriate for MAUI+Blazor apps. You should use the guidance in the other two articles mentioned ...

* https://learn.microsoft.com/aspnet/core/blazor/javascript-interoperability/call-javascript-from-dotnet
* https://learn.microsoft.com/aspnet/core/blazor/javascript-interoperability/call-dotnet-from-javascript

I'll make an additional adjustment to explicitly direct Blazor Hybrid readers to those two articles. Leave this issue open. It will close automatically when the PR merges.

**Comment by hawkerm** [comment]
Thanks @guardrex, I've found those articles as well, but haven't had luck applying any of the examples to my code for Maui Blazor Hybrid as well. Especially the module one that uses `IJSObjectReference`, as it throws an error that deserialization of interfaces isn't supported. Is there a dedicated example/sample repo somewhere that shows how to do this with Maui Blazor Hybrid?

**Comment by guardrex** [comment]
No, we don't ... and I've only assumed that the guidance works in Blazor Hybrid apps because the PU ("product unit") engineers never told me that it wouldn't work.

I think it best if I find some time to try a simple JS call and a simple .NET call in a MAUI-Blazor app to see if I can repro a problem. I have some time this morning, so I'll try a test within the hour and get back to you. *Stand-by .....................*

**Comment by guardrex** [comment]
I don't see a problem in a basic test. I used the first one, the `CallJs1` component example, which is in a nutshell ....

The script added in `index.html` under the Blazor script ...

```javascript
<script>
    window.convertArray = (win1251Array) => {
        var win1251decoder = new TextDecoder('windows-1251');
        var bytes = new Uint8Array(win1251Array);
        var decodedArray = win1251decoder.decode(bytes);
        console.log(decodedArray);
        return decodedArray;
    };
</script>
```

... with the code to call it in the `Home` component ...

```razor
@page "/"
@inject IJSRuntime JS

<h1>Hello, world!</h1>

Welcome to your new app.

<p>
    <button @onclick="ConvertArray">Convert Array</button>
</p>

<p>
    @text
</p>

@code {
    private MarkupString text;

    private uint[] quoteArray =
        new uint[]
        {
            60, 101, 109, 62, 67, 97, 110, 39, 116, 32, 115, 116, 111, 112, 32,
            116, 104, 101, 32, 115, 105, 103, 110, 97, 108, 44, 32, 77, 97,
            108, 46, 60, 47, 101, 109, 62, 32, 45, 32, 77, 114, 46, 32, 85, 110,
            105, 118, 101, 114, 115, 101, 10, 10,
        };

    private async Task ConvertArray()
    {
        text = new(await JS.InvokeAsync<string>("convertArray", quoteArray));
    }
}
```

I get ...

![image](https://github.com/dotnet/AspNetCore.Docs/assets/1622880/3101f4ca-80f3-49ee-ab92-676889498d19)

Next, I try a collocated module example `JsCollocation2` component at https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/?view=aspnetcore-8.0#load-a-script-from-an-external-javascript-file-js-collocated-with-a-component. That one goes ...

`Components/Pages/JsCollocation2.razor`:

```razor
@page "/js-collocation-2"
@implements IAsyncDisposable
@inject IJSRuntime JS

<PageTitle>JS Collocation 2</PageTitle>

<h1>JS Collocation Example 2</h1>

<button @onclick="ShowPrompt">Call showPrompt2</button>

@if (!string.IsNullOrEmpty(result))
{
    <p>
        Hello @result!
    </p>
}

@code {
    private IJSObjectReference? module;
    private string? result;

    protected async override Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            module = await JS.InvokeAsync<IJSObjectReference>("import",
                "./Components/Pages/JsCollocation2.razor.js");
        }
    }

    public async void ShowPrompt()
    {
        if (module is not null)
        {
            result = await module.InvokeAsync<string>(
                "showPrompt2", "What's your name?");
            StateHasChanged();
        }
    }

    async ValueTask IAsyncDisposable.DisposeAsync()
    {
        if (module is not null)
        {
            await module.DisposeAsync();
        }
    }
}
```

... added to the `NavMenu` to get to it ...

```razor
<div class="nav-item px-3">
    <NavLink class="nav-link" href="js-collocation-2">
        <span class="bi bi-list-nested-nav-menu" aria-hidden="true"></span> JS Col 2
    </NavLink>
</div>
```

... with ...

`Components/Pages/JsCollocation2.razor.js`:

```javascript
export function showPrompt2(message) {
  return prompt(message, 'Type your name here');
}
```

I have the prompt ...

![image](https://github.com/dotnet/AspNetCore.Docs/assets/1622880/189064f5-5354-4838-acd0-0caa57ffb4af)

... and the value is returned to the component ...

![image](https://github.com/dotnet/AspNetCore.Docs/assets/1622880/7968698f-c5e5-42c5-8451-43df4235f65a)

I can't repro a problem here.

I recommend that you open an issue for the product unit ... and you should place a minimal repro on GH for them to look at and/or show your exact code ...

https://github.com/dotnet/aspnetcore/issues

Please add ...

```
cc: @guardrex https://github.com/dotnet/AspNetCore.Docs/issues/31331
```

... to the bottom of your opening comment so that I can follow along. I might re-open this for doc work depending on what they say.
